python/ray/tune/trial_executor.py

```diff
@@ -237,17 +253,27 @@ def force_reconcilation_on_next_step_end(self) -> None:
         pass
 
     def _may_warn_insufficient_resources(self, all_trials):
-        if not any(trial.status == Trial.RUNNING for trial in all_trials):
+        # This is approximately saying we are not making progress.
+        if len(all_trials) == self._all_trials_size:
             if self._no_running_trials_since == -1:
                 self._no_running_trials_since = time.monotonic()
             elif time.monotonic(
             ) - self._no_running_trials_since > _get_warning_threshold():
-                # TODO(xwjiang): We should ideally output a more helpful msg.
+                if not is_ray_cluster():  # autoscaler not enabled
+                    cluster_resources = ray.cluster_resources()
```

Comment:
BTW, checking cluster_resources too often can cause major performance hits. We should be careful wherever we're making this call.

Reply:
Good call. Added caching.
For my own understanding, does this call involve network communications? Should head node already have information about cluster resources?